### DISCUSSION TITLE
Adding user grant with API

### DISCUSSION THREAD

**Discussion by aDogCalledSpot** [zitadel/zitadel#7520] [Q&A] [answered]
I'm trying to use the API to add a user grant and can't figure out what I'm doing wrong.

```
curl localhost:8100/management/v1/users/257055069126524930/grants -H "Authorization: Bearer $ADMIN_TOKEN" --json '{"projectId": "256331392655294467", "projectGrantId": "25706984691780813", "roleKeys": ["admin"] }' -v
```

returns

```
< HTTP/1.1 400 Bad Request
< Cache-Control: no-store
< Content-Type: application/json
< Expires: Fri, 08 Mar 2024 10:03:25 GMT
< Grpcgateway-Cache-Control: no-store
< Grpcgateway-Content-Type: application/grpc
< Grpcgateway-Pragma: no-cache
< Pragma: no-cache
< Vary: Origin
< X-Robots-Tag: none
< Date: Fri, 08 Mar 2024 10:03:25 GMT
< Content-Length: 170
<
* Connection #0 to host localhost left intact
{"code":9,"message":"Grant not found (COMMAND-4m9ff)","details":[{"@type":"type.googleapis.com/zitadel.v1.ErrorDetail","id":"COMMAND-4m9ff","message":"Grant not found"}]}
```

However, the grant exists:

```
curl localhost:8100/management/v1/projects/256331392655294467/grants/257345895153008643 -H "Authorization: $ADMIN_TOKEN" | jq
```
```json
{
  "projectGrant": {
    "grantId": "257345895153008643",
    "grantedOrgId": "257345895002013699",
    "grantedOrgName": "foo",
    "grantedRoleKeys": [
      "admin",
      "user"
    ],
    "state": "PROJECT_GRANT_STATE_ACTIVE",
    "projectId": "256331392655294467",
    "projectName": "Project1",
    "projectOwnerId": "256330751530827779",
    "projectOwnerName": "ZITADEL",
    "details": {
      "sequence": "29",
      "creationDate": "2024-03-08T08:21:46.659740Z",
      "changeDate": "2024-03-08T08:21:46.659740Z",
      "resourceOwner": "256330751530827779"
    }
  }
}
```

The creation works fine in the UI. So I'm not sure what the issue is.

**Comment by hifabienne** [comment]
I am not sure, but you are telling that you want to create but I can't see the POST in the curl example you sent. 
It should look something like this: 
`curl -L -X POST 'https://$CUSTOM-DOMAIN/management/v1/users/:userId/grants' \
-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>' \
--data-raw '{
  "projectId": "58949026806489455",
  "projectGrantId": "9847026806489455",
  "roleKeys": [
    "RoleKey1",
    "RoleKey2"
  ]
}'`

You can find the example here: https://zitadel.com/docs/apis/resources/mgmt/management-service-add-user-grant

**Reply by aDogCalledSpot** [reply]
Using --json automatically sets the method to POST

**Reply by hifabienne** [reply]
Ah missed that, I quickly checked the error message.
It looks like the project grant can't be found.
If you are sending the  "projectGrantId": "25706984691780813", this means that you need to have a project grant.
So ZITADEL checks if the pre conditions are met. So is the user in the same organization as the project?
Or did you grant the project to another organization and now you want to add a user grant in there?

If it is in the same organization the projectGrantId should be empty, otherwise make sure you have granted to project to the other org first

**Reply by aDogCalledSpot** [reply]
It's split up quite a bit:

- The administrator from which the bearer token is is in the ZITADEL org
- The project resides in OrgA
- The grant is for OrgB
- The user in question is in OrgC

This grant works in the UI. Is there maybe a way to see what request is being made there? I can't view the gRPC payloads.

**Reply by hifabienne** [reply, marked as answer]
Ok. So in that case you should have configurations like the following:
- ZITADEL
  - Service user
- OrgA
   - Project: 256331392655294467
      - Project Grant To OrgB 25706984691780813
- OrgB
   - Granted Project from Org A
   - User Grant you like to create here? for user from org C?
- OrgC
   - User: 257055069126524930

If the above is correct, you need to ensure that you are sending in the header x-zitadel-orgid header for the organizsatin B as your service user is not in the same organization

**Reply by aDogCalledSpot** [reply]
The header was what was missing, thanks.

I don't see a way to add the header in the Rust client API but I'll raise that issue in the corresponding repository. Thank you for your help :)
